src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyFlowGraph.cs

```diff
@@ -0,0 +1,197 @@
+            {
+                foreach (DependencyFlowEdge incomingEdge in node.IncomingEdges)
+                {
+                    // Don't use RemoveEdge as it assumes that this node will remain in the graph and
+                    // will cause modification of this IncomingEdges collection while iterating
+                    incomingEdge.From.OutgoingEdges.Remove(incomingEdge);
+                }
+                foreach (DependencyFlowEdge outgoingEdge in node.OutgoingEdges)
+                {
+                    // Don't use RemoveEdge as it assumes that this node will remain in the graph and
+                    // will cause modification of this OutgoingEdges collection while iterating
+                    DependencyFlowNode targetNode = outgoingEdge.To;
+                    outgoingEdge.To.IncomingEdges.Remove(outgoingEdge);
+                    // Recalculate the input channels
+                    targetNode.InputChannels = new HashSet<string>(targetNode.IncomingEdges.Select(e => e.Subscription.Channel.Name));
+                }
+            }
+        }
+
+        public void RemoveEdge(DependencyFlowEdge edge)
+        {
+            if (Edges.Remove(edge))
+            {
+                edge.From.OutgoingEdges.Remove(edge);
+                DependencyFlowNode targetNode = edge.To;
+                edge.To.IncomingEdges.Remove(edge);
+                // Recalculate the input channels
+                targetNode.InputChannels = new HashSet<string>(targetNode.IncomingEdges.Select(e => e.Subscription.Channel.Name));
+            }
+        }
+
+        /// <summary>
+        ///     Prune the graph, 
+        /// </summary>
+        /// <param name="graph"></param>
+        /// <param name="isInterestingNode"></param>
+        /// <param name="isInterestingEdge"></param>
+        public void PruneGraph(Func<DependencyFlowNode, bool> isInterestingNode,
+                               Func<DependencyFlowEdge, bool> isInterestingEdge)
+        {
+            // If nodes or edges are all interesting, then no reason to 
+            HashSet<DependencyFlowNode> unreachableNodes = new HashSet<DependencyFlowNode>(Nodes);
+            HashSet<DependencyFlowEdge> unreachableEdges = new HashSet<DependencyFlowEdge>(Edges);
+            Stack<DependencyFlowNode> nodes = new Stack<DependencyFlowNode>();
+
+            // Walk each root
+            foreach (DependencyFlowNode node in Nodes)
+            {
+                if (!isInterestingNode(node))
+                {
+                    continue;
+                }
+
+                nodes.Push(node);
+
+                while (nodes.Count != 0)
+                {
+                    DependencyFlowNode currentNode = nodes.Pop();
+
+                    if (!unreachableNodes.Contains(currentNode))
+                    {
+                        // Nothing to do
+                        continue;
+                    }
+                    unreachableNodes.Remove(currentNode);
```

Comment:
IIRC HashSet.Remove returns true if it removed something. You could avoid calling contains and just use the return value of remove.

Reply:
Nice, I like that pattern.